### PR TITLE
keyword elasticsearch

### PR DESCRIPTION
**Pull request by 363879721**

### CODE REVIEW

File: web/bfex/components/data_pipeline/tasks/keywords.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    connections.create_connection()
+    Keywords.init()
+
+    task = UpdateKeywordsFromScrape()
+    search = Faculty.search()
+
+    results = search.query('match', name="Osmar.Zaiane")
+    task.run(results)
+ 
+    ##search = Keywords.search()
```
Comment (MandyMeindersma) [inline web/bfex/components/data_pipeline/tasks/keywords.py]: plz delete comments

File: web/bfex/components/data_pipeline/tasks/scrape.py [hunk 2]
```diff
@@ -39,18 +40,24 @@ def run(self, data):
...
             scraper = ScraperFactory.create_scraper(faculty_directory_url, ScraperType.PROFILE)
             scrapp = scraper.get_scrapps()[0]
-            tuple = (faculty_name, scrapp)
+
+            tuple = (faculty_name,scrapp)
             tuplelist.append(tuple)
             print(tuple)
+        
 
+        #keygen = RakeApproach()
```
Comment (MandyMeindersma) [inline web/bfex/components/data_pipeline/tasks/scrape.py]: plz no comments

File: web/bfex/components/data_pipeline/tasks/scrape.py [hunk 3]
```diff
@@ -39,18 +40,24 @@ def run(self, data):
...
         return tuplelist
 
 
 if __name__ == "__main__":
     task = FacultyPageScrape()
 
-    if task.is_requirement_satisfied("JNelson.Amaral"):
-        task.run("JNelson.Amaral")
+    if task.is_requirement_satisfied("Stan.Boutin"):
+        task.run("Stan.Boutin")
```
Comment (MandyMeindersma) [inline web/bfex/components/data_pipeline/tasks/scrape.py]: why is this here?
I forget

File: web/bfex/components/data_pipeline/tasks/keywords.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                not isinstance(data[0], str) or
+                not isinstance(data[1], Scrapp)):
+            satisfied = False
+
+        return satisfied
+
+
+    def run(self,data):
+       # professor_id = data[0]
+        #scrapp = data[1]
```
Comment (MandyMeindersma) [inline web/bfex/components/data_pipeline/tasks/keywords.py]: comments?

File: web/bfex/components/data_pipeline/tasks/keywords.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        for faculty in data:
+            print(faculty.faculty_id)
```
Comment (MandyMeindersma) [inline web/bfex/components/data_pipeline/tasks/keywords.py]: we should be printing something every time this runs but should we be printing this all?

File: web/bfex/components/data_pipeline/tasks/scrape.py (hunk 3 above)
Comment (GarryBullock) [reply web/bfex/components/data_pipeline/tasks/scrape.py]: Yeah, I think I have some of these leftover as well, but lets move them to tests.

File: web/bfex/components/key_generation/rake_approach.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+from bfex.components.key_generation.key_generation_approach import KeyGenerationApproach
+from bfex.components.key_generation.RAKE import rake
+from bfex.components.scraper.scrapp import Scrapp
+import io
+
+
+class RakeApproach(KeyGenerationApproach):
+    def __init__(self):
+        self.description = """ rake """
```
Comment (GarryBullock) [inline web/bfex/components/key_generation/rake_approach.py]: Can we get a description of how this approach works? This could be useful if we want to display that in the UI, or just providing further context.

File: web/bfex/components/data_pipeline/tasks/persist.py [hunk 7]
```diff
@@ -51,6 +51,9 @@ def run(self, data):
             if "researchid_link" in scrapp.meta_data:
                 faculty.research_id = scrapp.meta_data["researchid_link"]
 
+            if "text" in scrapp.meta_data:
+                 faculty.text = scrapp.meta_data["text"]
```
Comment (GarryBullock) [inline web/bfex/components/data_pipeline/tasks/persist.py]: I'm not sure if we want to store this data inside the faculty member document. It makes sense in that it is obviously linked to a single person, and fits in better with our current workflow. But I also think it'd be nice to have all documents in a seperate index. There are certain operations (ie. tf-idf) that we may not want to run on everything in the faculty index.

Let's revisit this during our next meeting.

File: web/bfex/components/data_pipeline/tasks/keywords.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            professor_id = faculty.faculty_id
+            print(professor_id)
+
+        search_results = Faculty.search().query('match', faculty_id=professor_id).execute()
```
Comment (GarryBullock) [inline web/bfex/components/data_pipeline/tasks/keywords.py]: Searching by the id is different than searching for text. You can use the safe_get method defined in models to do this in a cleaner way. No sense searching through all faculty when elastic can easily get the one we want based on the unique id. Take a look at the data ingestion blueprint for an example.

File: web/bfex/components/data_pipeline/tasks/keywords.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        faculty = search_results[0]
+        keywords = []
+        if faculty.text:
+            text = faculty.text
+
+            keygen = RakeApproach()
+
+            rake_keyword = keygen.generate_keywords(text)
+
+            get_keyword = Keywords.search().query('match', faculty_id = professor_id).execute()
```
Comment (GarryBullock) [inline web/bfex/components/data_pipeline/tasks/keywords.py]: Note above comment about searching for a specific professor.

File: web/bfex/components/data_pipeline/tasks/scrape.py [hunk 10]
```diff
@@ -3,31 +3,32 @@
 from bfex.models import Faculty
 from bfex.common.utils import URLs, FacultyNames
 from bfex.components.data_pipeline.tasks.task import Task
-
+from bfex.components.key_generation.rake_approach import *
+from bfex.components.data_pipeline.tasks.keywords import UpdateKeywordsFromScrape
```
Comment (GarryBullock) [inline web/bfex/components/data_pipeline/tasks/scrape.py]: If you add keywords to the __init__.py of bfex/components/data_pipeline/tasks, it makes the imports a little nicer, but thats not necessary.

File: web/bfex/components/data_pipeline/tasks/scrape.py [hunk 11]
```diff
@@ -3,31 +3,32 @@
...
 
     def is_requirement_satisfied(self, data):
-        """Checks the requirements for a faculty page scraping are satisfied.
+        """ Checks the requirements for a faculty page scraping are satisfied.
 
         For a Faculty Directory page scrape, we can get the data in one of 2 ways. Either a faculty member name is
         passed to us, or an instance of a faculty member is passed in.
-        :param data: str or Faculty instance.
-        :return: True if the name is a valid one, else False
+        :param data:
```
Comment (GarryBullock) [inline web/bfex/components/data_pipeline/tasks/scrape.py]: So what does this work on? What does it take in?

File: web/bfex/components/data_pipeline/tasks/scrape.py [hunk 12]
```diff
@@ -3,31 +3,32 @@
...
+        # Should this return the data as it is expected for running the task?
+        if isinstance(data, str):
+            # Apply any additional checks or validations that need to occur on the data
+            return FacultyNames.validate_name(data)
 
-        return isinstance(data, list)
+        if isinstance(data, Faculty):
+            return FacultyNames.validate_name(data.name)
 
     def run(self, data):
```
Comment (GarryBullock) [inline web/bfex/components/data_pipeline/tasks/scrape.py]: Lets make sure to add a docstring that describes what is happening here.

**Comment by GarryBullock** [comment]
@363879721 Can you clarify your help wanted? What do you need help with?
